src/xdsl/immutable_ir.py

```diff
@@ -0,0 +1,662 @@
+                    new_operands.append(env[operand])
+                    substitution_required = True
+                else:
+                    new_operands.append(operand)
+
+            # If any updates to this op are required we rebuild it
+            if substitution_required:
+                substituted_op = from_op(op,
+                                         operands=new_operands,
+                                         regions=new_regions,
+                                         env=env)
+                assert len(substituted_op) == 1
+                return substituted_op[-1]
+            return op
+
+        ops = [substitute_if_required(op) for op in ops]
+
+        return cls(args=block_args, ops=ops)
+
+    @classmethod
+    def from_mutable(
+        cls,
+        block: Block,
+        value_map: Optional[Dict[SSAValue, ISSAValue]] = None,
+        block_map: Optional[Dict[Block, IBlock]] = None,
+    ) -> IBlock:
+        if value_map is None:
+            value_map = {}
+        if block_map is None:
+            block_map = {}
+
+        args: List[IBlockArg] = []
+        for arg in block.args:
+            immutable_arg = IBlockArg(arg.typ, None, arg.index)  # type: ignore
+            args.append(immutable_arg)
+            value_map[arg] = immutable_arg
+
+        immutable_ops = [
+            IOp.from_mutable(op, value_map=value_map, block_map=block_map)
+            for op in block.ops
+        ]
+
+        return IBlock(args, immutable_ops)
+
+    def get_mutable_copy(
+            self,
+            value_mapping: Optional[Dict[ISSAValue, SSAValue]] = None,
+            block_mapping: Optional[Dict[IBlock, Block]] = None) -> Block:
+        if value_mapping is None:
+            value_mapping = {}
+        if block_mapping is None:
+            block_mapping = {}
+
+        new_block = Block.from_arg_types(self.arg_types)
+        for idx, arg in enumerate(self.args):
+            value_mapping[arg] = new_block.args[idx]
+        block_mapping[self] = new_block
+
+        for immutable_op in self.ops:
+            new_block.add_op(
+                immutable_op.get_mutable_copy(value_mapping=value_mapping,
+                                              block_mapping=block_mapping))
+        return new_block
+
+    def walk(self, fun: Callable[[IOp], None]) -> None:
+        for op in self.ops:
+            op.walk(fun)
+
+
+def get_immutable_copy(op: Operation) -> IOp:
+    return IOp.from_mutable(op, {})
+
+
+@dataclass(frozen=True)
+class OpData:
+    name: str
+    op_type: type[Operation]
+    attributes: Dict[str, Attribute]
+
+
+@dataclass(frozen=True)
+class IOp:
+    __match_args__ = ("op_type", "operands", "results", "successors",
+                      "regions")
+    _op_data: OpData
+    operands: IList[ISSAValue]
+    results: IList[IResult]
+    successors: IList[IBlock]
+    regions: IList[IRegion]
+
+    def __init__(self, op_data: OpData, operands: List[ISSAValue],
```

Comment:
You can make the type of the input a `Seq[ISSAValue]` rather than a `List[ISSAValue]`.
The `Seq` type will allow anything list-like (e.g. a `FrozenList`) to be passed in and the `Seq` type also does not explose the mutable API, so that any attempt to modify the value will result in a type error.